src/common/hist_util.h

```diff
@@ -19,6 +19,70 @@
 namespace xgboost {
 namespace common {
 
+template<typename T>
+struct SimpleArray {
+  ~SimpleArray() {
+    free(ptr_);
+    ptr_ = nullptr;
+  }
+
+  void resize(size_t n) {
+    T* ptr = static_cast<T*>(malloc(n*sizeof(T)));
+    memcpy(ptr, ptr_, n_ * sizeof(T));
+    free(ptr_);
+    ptr_ = ptr;
+    n_ = n;
+  }
+
+  T& operator[](size_t idx) {
```

Comment:
What is the advantage of using `SimpleArray` over `std::vector`? We still have one extra indirection each time we index the array.

Reply:
I replaced std::vector by SimpleArray because std::vector has to slow resize() function, it fill memory by very not performance manner.
Amortization of allocation/deallocation is not needed here, because I need to call resize() only one time.